File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def helper(root,left,right):
            if not root:
                return True
            if not (left < root.val < right):
                return False
            
            right_ans = helper(root.right ,root.val,right)
            left_ans = helper(root.left ,left,root.val)

            return right_ans and left_ans
        return helper(root,float('-inf'),float('inf'))

        

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root==null){
            return "";
        }
        StringBuilder sb = new StringBuilder();
        Queue<TreeNode> q = new LinkedList<>();
        
        q.offer(root);

        while(!q.isEmpty()){
            TreeNode node = q.poll();
        
            if (node==null){
                sb.append("#,");
                continue;
            }
            sb.append(node.val).append(",");
            q.offer(node.left);
            q.offer(node.right);

        }
        System.out.print(sb);
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.isEmpty()){
            return null;
        }
        String arr[] = data.split(",");
        Queue<TreeNode> q = new LinkedList<>();
        TreeNode root = new TreeNode(Integer.parseInt(arr[0]));
        q.offer(root);
        int i = 1;

        while (i<arr.length && !q.isEmpty()){
            TreeNode node = q.poll();

            if (!arr[i].equals("#")){
                node.left =  new TreeNode(Integer.parseInt(arr[i]));
                q.offer(node.left);
            }
            i++;
            if (!arr[i].equals("#")){
                node.right =  new TreeNode(Integer.parseInt(arr[i]));
                q.offer(node.right);
            } 
            i++;
        }
        
        return root;
    }
}
